File: javascript/isbn-verifier/isbn-verifier.js

export const isValid = (isbn) => {
    let count = 0;
    let sum = 0;
    for (let i = 0; i < isbn.length; i++) {
        let char = isbn[i];
        if (char === '-') {
            continue;
        }
        if (char === 'X') {
            if (count !== 9) {
                return false;
            }
            sum += 10;
            count++;
            break;
        }
        char = char.charCodeAt(0) - 48;
        if (char < 0 || char > 9) {
            return false
        }
        sum += char * (10 - count);
        count++;
    }
    return count === 10 && sum % 11 === 0
};

File: zig/isbn-verifier/isbn_verifier.zig
pub fn isValidIsbn10(s: []const u8) bool {
    var count: u8 = 0;
    var sum: u16 = 0;
    for (s) |c| {
        if (c >= '0' and c <= '9') {
            sum += (c - '0') * (10 - count);
            count += 1;
            continue;
        }
        if (c == 'X' and count == 9) {
            sum += 10;
            count += 1;
            continue;
        }
        if (c != '-') {
            return false;
        }
    }
    return count == 10 and sum % 11 == 0;
}
